File: wander/src/commonMain/kotlin/dev/ligature/wander/Wander.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature.wander

import arrow.core.Either
import arrow.core.flatMap
import dev.ligature.LigatureError
import dev.ligature.wander.interpreter.Bindings
import dev.ligature.wander.interpreter.eval
import dev.ligature.wander.lexer.Token
import dev.ligature.wander.lexer.tokenize
import dev.ligature.wander.library.common
import dev.ligature.wander.model.Element
import dev.ligature.wander.parser.parse

interface WanderError : LigatureError

suspend fun run(script: String, bindings: Bindings = common()): Either<WanderError, Element> =
    tokenize(script).flatMap { tokens: List<Token> -> parse(tokens) }.flatMap { eval(it, bindings) }

File: wander/shared/src/main/scala/dev/ligature/wander/command/Wander.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.wander.command

import dev.ligature.Identifier

def run(script: String): Either[WanderError, WanderValue] =
  for
    tokens      <- tokenize(script)
    commands    <- parse(tokens)
    result      <- eval(commands)
  yield result

def runPrint(script: String): String =
  run(script) match
    case Right(value) => printWanderValue(value)
    case Left(err) => err.message

case class WanderError(message: String)

enum WanderValue:
    case Integer(value: Long)
    case StringLiteral(value: String)
    case BooleanLiteral(value: Boolean)
    case Identifier(value: String)
    case Tuple(values: Seq[WanderValue])
    case Nothing

def printWanderValue(value: WanderValue): String =
  value match
    case WanderValue.Integer(value) => value.toString()
    case WanderValue.StringLiteral(value) => s"\"$value\""
    case WanderValue.Nothing => "nothing"
    case WanderValue.BooleanLiteral(value) => value.toString()
    case WanderValue.Identifier(value) => s"<$value>"
    case WanderValue.Tuple(values) => ???